pages/management/commands/upload_docx.py

```diff
@@ -0,0 +1,138 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+#    this list of conditions and the following disclaimer.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+Command to extract text from either an uploaded document
+or a Google Docs link to set as a PageElement's text.
+"""
+import logging
+
+import os
+import re
+import zipfile
+from io import BytesIO
+from xml.etree.ElementTree import XML, ParseError
+
+import requests
+from django.core.management.base import BaseCommand, CommandError
+
+from ...models import PageElement
+
+LOGGER = logging.getLogger(__name__)
+
+
+WORD_NAMESPACE = '{http://schemas.openxmlformats.org/wordprocessingml/2006/main}'
+PARA = WORD_NAMESPACE + 'p'
+TEXT = WORD_NAMESPACE + 't'
+
+class Command(BaseCommand):
+    help = 'Uploads a .docx file from a local system or Google Drive link'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            'source',
+            type=str,
+            help='Source ("upload" for local file, "drive" for Google Drive link)')
+        parser.add_argument(
+            'identifier',
+            type=str,
+            help='File path or Google Drive link')
+        parser.add_argument(
+            'page_element_slug',
+            type=str,
+            help='Slug of the PageElement to update')
+
+    @staticmethod
+    def format_drive_url(url):
+        if 'docs.google.com/document' not in url:
+            raise CommandError(f'Invalid Google Docs URL: "{url}"')
+
+        match = re.search(r'/d/([0-9A-Za-z_-]{28,})/', url)
+        if not match:
+            raise CommandError(f'Could not extract document ID from Google Docs URL: "{url}"')
+
+        doc_id = match.group(1)
+        return f'https://docs.google.com/document/d/{doc_id}/export?format=docx'
+
+    @staticmethod
+    def get_docx_text(file_obj):
```

Comment:
I would have expected the use of a library like [python-docx](https://python-docx.readthedocs.io/en/latest/), especially because more documents than none have embed pictures or tables, but that works as a first pass.